### ISSUE TITLE
[frontend] Kubeflow does N queries for in the "Runs" page

### ISSUE THREAD

**Issue by asaff1** [kubeflow/pipelines#11346]
### Environment

*  How did you deploy Kubeflow Pipelines (KFP)?
Kubeflow pipelines standalone, AWS setup

### Steps to reproduce
Open your network panel in dev tools.  Navigate to the Runs page and look at your network console. Try to increase page size in the UI, and see many requests.

When navigating to the "Runs" page, kubeflow will send an API call to fetch a list of runs. There are two problems here:
1) This query returns a `total_size` field, which does unneeded COUNT(*) query on the whole "run_details" table, the count is not even displayed in the UI.
2) More important, after the run list is fetched, the UI will do an **API call per run** (which is a DB query) to get its associated pipeline.  This is very slow. Instead, the runs API could simply do an SQL JOIN to get the pipeline info.
This is really slow, even for page size of 10, tested with my medium size RDS instance. When page size is 100, this page will do **over 100 SQL queries**.

![image](https://github.com/user-attachments/assets/3c441581-5467-4285-8ac3-855e41a91a55)

### Expected result
To get the runs data with the pipeline info, one query should be enough. Page should load faster.
### Materials and Reference

---

Impacted by this bug? Give it a 👍.

**Comment by droctothorpe** [comment]
https://github.com/kubeflow/pipelines/pull/10797 should resolve this. Make sure that you're on the latest version of KFP.

**Comment by asaff1** [comment]
@droctothorpe I've tried to replace only the `image:` in the ml-pipeline-ui deployment. I changed from `gcr.io/ml-pipeline/frontend:2.1.0` to `gcr.io/ml-pipeline/frontend:2.3.0`. And still, the issue exists.. I do see the PR in the changelog of 2.3.0, so it is really strange..

**Comment by droctothorpe** [comment]
That's odd. Maybe look at the source code in the chrome console and make sure it includes the code in question. Alternatively, you can try to build and push a new image from the master branch and deploy that. Another option is to run the UI in development mode from your local machine. You can see how to do that [here](https://github.com/kubeflow/pipelines/blob/master/frontend/README.md#proxy-to-a-real-cluster).

**Comment by asaff1** [comment]
@droctothorpe I believe that the build of 2.3.0 didn't build your PR. Would be great if you can try it yourself, test the `gcr.io/ml-pipeline/frontend:2.3.0` image.

**Comment by rimolive** [comment]
@asaff1 Can you confirm your issue is solved if you build the image manually and use in your deployment? If the latest image did not fix this issue, we need to reopen it.

**Comment by asaff1** [comment]
@rimolive I will try to build and will update. What I can say for sure is that the 2.3.0 image still have the bug. This doesn't match the release notes of the 2.3.0.

**Comment by droctothorpe** [comment]
2.4 was just released, @asaff1. Have you tried against that version of the frontend?

**Comment by droctothorpe** [comment]
Crud. Guess you'll need to build yourself until that's cut.

**Comment by HumairAK** [comment]
@asaff1 in 2.4 we have moved away from GCR and instead are using GHCR: https://github.com/kubeflow/pipelines/pkgs/container/kfp-frontend

**Comment by asaff1** [comment]
@HumairAK 
I see. I've tried `ghcr.io/kubeflow/kfp-frontend:2.4.0`, still in the frontend in the "Runs" page it does many queries to get associated pipeline per run. The issue is not solved.

**Comment by droctothorpe** [comment]
Thanks for confirming, @asaff1. I'll take a look at this.

**Comment by droctothorpe** [comment]
@asaff1, I just tested on my local machine. I have 2.3 deployed on a local (colima-backed) k8s cluster. The image is set to `gcr.io/ml-pipeline/frontend:2.3.0`. It's only making a single request to the backend even though there are 10 distinct runs. Here's a screenshot as evidence:

<img width="1398" alt="Image" src="https://github.com/user-attachments/assets/26f78fe5-d094-460a-9c14-9dd0b64c65a9" />

The non-duplicated request being made is: `http://localhost:3000/apis/v2beta1/runs?page_token=&page_size=10&sort_by=created_at%20desc&filter=%257B%2522predicates%2522%253A%255B%257B%2522key%2522%253A%2522storage_state%2522%252C%2522operation%2522%253A%2522NOT_EQUALS%2522%252C%2522string_value%2522%253A%2522ARCHIVED%2522%257D%255D%257D`.

**Comment by droctothorpe** [comment]
> [@droctothorpe](https://github.com/droctothorpe) I see. BTW, same issue for the experiments page - it loads the last 5 runs times the number of experiments shown. I'm using kubeflow 2.1.0 and not latest 2.3.0 because of regressions that were introduced (probably when argo was upgraded - for example retry doesn't work)

And here's the experiments page only making a single request to the runs API:

<img width="1395" alt="Image" src="https://github.com/user-attachments/assets/4dd27e94-d4b5-4fcb-9900-49ec1b8d5d37" />

**Comment by droctothorpe** [comment]
To rule out the possibility that multiple experiments are the problem, I tested that too. Still just one runs request:


<img width="1418" alt="Image" src="https://github.com/user-attachments/assets/044be516-f978-4380-9a3e-ee96ba84dd9e" />

**Comment by asaff1** [comment]
@droctothorpe Looks different on my end - In your screenshot of the "Runs" page I see the [View pipeline] placeholder, on my 2.3.0 UI it shows the actual pipeline name. Is it possible it is due to a configuration difference passed to the UI deployment, or something different the UI gets from the backend?

I only tried to upgrade the UI image to 2.3.0 (and 2.4.0), I didn't upgrade the API backend (ml-pipeline), and definitely not the whole kubeflow installation. My backend (and the kubeflow release) is still at 2.1.0. I assumed this issue is only a UI bug.

**Comment by droctothorpe** [comment]
I recommend checking if you can reproduce the error against a full KFP 2.3.0 install.

**Comment by asaff1** [comment]
@droctothorpe I cannot fully install 2.3.0, because it had another bugs. (I checked full kubeflow installations from 2.1.0 to 2.4.0 and settled on 2.1.0).
I don't understand why I cannot upgrade the UI alone? as I see in your fix PR, it only modifies the UI.

**Comment by droctothorpe** [comment]
@asaff1 The PR was tested against and contributed as a fix for the current version of the code, not older versions and not environments that mix and match different versions. You can try upgrading just your KFP backend version or diffing the relevant code between 2.1.0 and the tip of master to identify the discrepancy.

**Comment by droctothorpe** [comment]
The screenshot in [this](https://github.com/kubeflow/pipelines/issues/11346#issuecomment-2608774151) comment shows multiple manually triggered runs that are not part of recurring runs and the issue is not present. How are you triggering your runs? In the linked example, I used `create_run_from_pipeline_func`.

**Comment by asaff1** [comment]
@droctothorpe can you explain when exactly the [View pipeline] is showed and when the actual pipeline name is showing?
Honestly I do prefer seeing the pipeline name. It is easier when I have one experiment that runs different types of pipelines.
I should also say, I'm running kfp v1 pipelines (using `ContainerOp`) - can that be related?

**Comment by droctothorpe** [comment]
I suspect `[View pipeline]` is shown when you create a run using the `Create run` button, since it links to a specific pipeline. All of our end users create runs from the SDK using `create_run_from_pipeline_func/package`, which doesn't link runs to specific pipeline definitions. Hopefully you can use this pattern. That being said, I think you have identified a legitimate gap in the UI (i.e. redundant requests when listing runs created from the UI). How would you feel about addressing that gap? It probably involves mimicking some of the changes in [this](https://github.com/kubeflow/pipelines/pull/10797/files#diff-99290b15e13174d7f604f4772ef2a5487368fa3b5800c9197b5a30bb4ed91e39) PR in whatever branch of the logic handles the `[View pipelines]` scenario.

**Comment by droctothorpe** [comment]
I haven't looked at this code in a bit so take this with a grain of salt, but I think that [this](https://github.com/droctothorpe/pipelines/blob/cffecf8424b97dc336d04fa336d3b7463c1a58d4/frontend/src/pages/RunList.tsx#L491) function is what gets called n times for n runs. My guess is that for SDK-created runs, [this](https://github.com/droctothorpe/pipelines/blob/cffecf8424b97dc336d04fa336d3b7463c1a58d4/frontend/src/pages/RunList.tsx#L494) condition is not met, so we never make the individual requests / lookups. There's a possibility that the batch runs lookup does not contain the pipeline version info, in which case, there may not be a way to circumvent the individual lookups, other than eliminating the `[View pipeline]` functionality altogether. If you want to go down that route, I recommend getting design approval on a WG call or something to that effect before implementing it.

**Comment by asaff1** [comment]
@droctothorpe I believe it is in fact a backend issue. (Not just a UI issue).
The runs API doesn't return the experiment and pipeline names, and thus the UI has to do extra API calls to get the experiment name and pipeline name (and version name) for each run. That means, for N runs in the list the UI will actually do additional 2*N API calls (SQL queries) = N experiments + N pipelines. This is what I observe.

You can see in the following runs API response, it only contains, `"pipeline_version_reference"` and `"experiment_id"`. But not names for these. So the UI does extra calls to get them.

The solution, is that the backend needs to do a simple SQL join to fetch pipeline name + experiment name in the runs API. As I see this is possible by a `JOIN` with the "resource_references" table in the DB.

![Image](https://github.com/user-attachments/assets/333e85fa-7766-4bdb-810c-4ef780f31e00)

**Comment by droctothorpe** [comment]
That makes sense to me, @asaff1. Updating the runs endpoint to (optionally) handle the join would offload the burden from the UI. Better to fix things in the backend than the frontend.

**Comment by droctothorpe** [comment]
It's going to involve modifying both the backend and the frontend. I think opening a standalone issue to get approval on the backend change (the optional join) first probably makes sense. You can probably expedite things by bringing it up during a WG meeting. There's one tomorrow at 1PM EST. Once you get approval, I recommend submitting a PR with the backend changes only, then a follow-up PR with the frontend changes that take advantage of the updated backend.

**Comment by asaff1** [comment]
@droctothorpe I've opened a backend issue
https://github.com/kubeflow/pipelines/issues/11867

**Comment by dlemfh** [comment]
Also, I noticed that frontend code calls listRuns() and listExperiments() sequentially.
![Image](https://github.com/user-attachments/assets/893ab251-214e-4cde-87d3-3588a8b3fd9c)

I think more speed up could be possible if the two are called in parallel. 🤗

**Comment by droctothorpe** [comment]
Apologies for the delay, @dlemfh. I was out on paternity leave. I think both of those are good suggestions. Would you be amenable to contributing them? I'd be happy to review them.

**Comment by dlemfh** [comment]
No worries @droctothorpe!

But sorry I've stopped working on Kubeflow for a while and I don't think I can find time for it any time soon.

Hoping anyone interested can pick up from here! ☺️

**Comment by asaff1** [comment]
@droctothorpe is this issue is scheduled to be resolved? (And #11867)

**Comment by droctothorpe** [comment]
It's an open issue and we intend to address it but it's not explicitly scheduled on the road map yet. If you want to tackle it to move it up the queue, we would appreciate the contribution and I'd be happy to review the PR.

**Comment by Sharkyii** [comment]
Wanted to check something before starting, since it changes the plan a bit.

There's already a PR for half of this, #12827 from @puwun . It adds experiment_display_name and pipeline_version_display_name straight onto the Run proto with LEFT JOINs in run_store.go, which is the real fix for the pipeline and experiment lookups, an actual join instead of batching things client side. It's stuck though. Never got an /ok-to-test, only a bot ever reviewed it, and it's been stale since June. It also says the frontend part, actually using those fields in RunList.tsx instead of the current per-run getPipelineVersion calls, was meant to be a follow-up PR that never got opened.

One thing neither #12827 nor #11867 touches is the other half of this bug: ListRuns in run_store.go still runs an unconditional total_size query every call, and the frontend never reads that value at all. That part is completely untouched right now.

So rather than redo #12827's backend work, I'd rather start with the total_size fix since it's independent and won't collide with anything. For the pipeline lookup half, it'd make more sense to help get #12827 unstuck, rebase it, reopen #11867, ping for /ok-to-test, and maybe pick up the missing frontend follow-up, than build a second fix for the same thing. Let me know if @puwun  is still around or if it's fair game to help move forward.

**Comment by rvolchek** [comment]
🏁 flagged for review

**Comment by jeffspahr** [comment]
PR #13951 addresses part of this issue:

- It adds an opt-in `skip_count` parameter and uses it in frontend run-list requests that do not consume `total_size`, avoiding the unused count query on those requests.
- It deduplicates identical pipeline-version point lookups. For `R` displayed runs referencing `U` unique `(pipelineId, pipelineVersionId)` pairs, this reduces version requests from `R` to `U`.

It does **not** eliminate the general pipeline-version N+1 behavior. If every displayed run references a different version, then `U = R` and the number of version requests is unchanged. The backend batch/join solution requested here therefore remains open work, and this issue should stay open after #13951 merges.
